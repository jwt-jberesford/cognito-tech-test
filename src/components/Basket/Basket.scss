@import "../../variables.scss";

.styled-basket {
  min-height: 100vh;
  height: 100%;
  // width: 30rem;
  width: 100%;
  background: $light-primary;
  // border-left: 3px solid $green-main;
  position: fixed;
  top: 0;
  right: 0;
  z-index: 1;
  padding: 2rem;
  display: flex;
  flex-direction: column;

  h3 {
    font-weight: 500;
    font-size: 1.5rem;
    margin: 1.7rem 0;
    text-decoration: underline;
    text-underline-offset: 4px;
    text-decoration-thickness: 2px;
    text-decoration-color: $green-main;
  }

  .checkout-button-wrapper {
    margin-top: auto;
    button {
      width: 100%;
    }
  }
}

.dark {
  background: $dark-primary;
}

.basket-items {
  overflow-y: auto;
  margin: 1rem 0;
  flex-grow: 1;
  position: relative;

  &::-webkit-scrollbar {
    width: 0.5rem;
    background: light;
  }

  &::-webkit-scrollbar-track {
    background-color: transparent;
  }

  &::-webkit-scrollbar-thumb {
    background-color: $green-main;
    border-radius: 10rem;

    &:hover {
      background-color: darken($green-main, 10%);
    }
  }
}

.break-line {
  width: 100%;
  height: 2px;
  background-color: rgb(240, 240, 240);
}

.basket-empty {
  display: flex;
  justify-content: center;
  align-items: center;
  flex-grow: 1;

  span {
    font-size: 1.2rem;
  }
}

@media (min-width: 700px) {
  .styled-basket {
    width: 25rem;
    border-left: 3px solid $green-main;
  }
}

@media (min-width: 900px) {
  .styled-basket {
    width: 30rem;
    border-left: 3px solid $green-main;
  }
}
