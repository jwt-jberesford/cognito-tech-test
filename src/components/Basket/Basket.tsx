import React from "react";
import { motion, AnimatePresence } from "framer-motion";
import { useSelector } from "react-redux";
import { RootState } from "../../store/store";
import "./Basket.scss";
import BasketItem from "../BasketItem/BasketItem";
import Button from "../Button/Button";

const Basket: React.FC = () => {
  const basketActive = useSelector((state: RootState) => state.basket.active);
  const basketItems = useSelector((state: RootState) => state.basket.items);

  return (
    <AnimatePresence>
      {basketActive && (
        <motion.div
          className="styled-basket"
          initial="hidden"
          animate="visible"
          exit="exit"
          transition={{ duration: 0.5, ease: "easeInOut" }}
        >
          <h3>Your Basket</h3>

          {basketItems.length ? (
            <>
              <div className="basket-items">
                {basketItems.map((item) => (
                  <React.Fragment key={item.id}>
                    <BasketItem product={item} />
                    <div className="break-line" />
                  </React.Fragment>
                ))}
              </div>
              <div className="checkout-button-wrapper">
                <Button text="Go to checkout" primary={true} size="large" />
              </div>
            </>
          ) : (
            <div className="basket-empty">
              <span>Add some items to the basket to proceed</span>
            </div>
          )}
        </motion.div>
      )}
    </AnimatePresence>
  );
};

export default Basket;
