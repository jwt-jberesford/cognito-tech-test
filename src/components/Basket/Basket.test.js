import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import Basket from "./Basket";
import "@testing-library/jest-dom";

describe("Basket component unit tests", () => {
  const mockStore = configureStore();
  let store;

  beforeEach(() => {
    store = mockStore({
      basket: {
        active: false,
        items: [],
      },
      theme: {
        theme: "light",
      },
    });
  });

  test("Basket does not render when state inactive", () => {
    render(
      <Provider store={store}>
        <Basket />
      </Provider>
    );

    expect(screen.queryByText("basket")).toBeNull();
  });

  test("Basket will render items", () => {
    store = mockStore({
      basket: {
        active: true,
        items: [{ id: 1, name: "apple" }],
      },
      theme: {
        theme: "light",
      },
    });

    render(
      <Provider store={store}>
        <Basket />
      </Provider>
    );

    expect(screen.getByText("apple")).toBeInTheDocument();
  });

  test("Basket renders correctly when active", () => {
    store = mockStore({
      basket: {
        active: true,
        items: [{ id: 1, name: "apple" }],
      },
      theme: {
        theme: "light",
      },
    });

    render(
      <Provider store={store}>
        <Basket />
      </Provider>
    );

    expect(screen.getByText("Your Basket")).toBeInTheDocument();

    const basketHeading = screen.queryByText("Your Basket");
    expect(basketHeading).toBeInTheDocument();
  });

  test("Basket updates when items are added", () => {
    store = mockStore({
      basket: {
        active: true,
        items: [],
      },
      theme: {
        theme: "light",
      },
    });

    const { rerender } = render(
      <Provider store={store}>
        <Basket />
      </Provider>
    );

    expect(
      screen.queryByText("Add some items to the basket to proceed")
    ).toBeInTheDocument();

    store = mockStore({
      basket: {
        active: true,
        items: [{ id: 1, name: "Apple", price: 1.99, quantity: 1 }],
      },
      theme: {
        theme: "light",
      },
    });

    rerender(
      <Provider store={store}>
        <Basket />
      </Provider>
    );

    expect(screen.getByText("Apple")).toBeInTheDocument();
  });

  test("Basket updates when items are removed", () => {
    store = mockStore({
      basket: {
        active: true,
        items: [{ id: 1, name: "Apple", price: 1.99, quantity: 1 }],
      },
      theme: {
        theme: "light",
      },
    });

    const { rerender } = render(
      <Provider store={store}>
        <Basket />
      </Provider>
    );

    expect(screen.getByText("Apple")).toBeInTheDocument();

    store = mockStore({
      basket: {
        active: true,
        items: [],
      },
      theme: {
        theme: "light",
      },
    });

    rerender(
      <Provider store={store}>
        <Basket />
      </Provider>
    );

    // Expect the item to no longer be in the document
    expect(screen.queryByText("Apple")).toBeNull();
  });

  test("Empty basket message appears when no items are present", () => {
    store = mockStore({
      basket: {
        active: true,
        items: [],
      },
      theme: {
        theme: "light",
      },
    });

    render(
      <Provider store={store}>
        <Basket />
      </Provider>
    );

    expect(
      screen.getByText("Add some items to the basket to proceed")
    ).toBeInTheDocument();
  });

  test("Basket component applies the theme correctly", () => {
    store = mockStore({
      basket: {
        active: true,
        items: [{ id: 1, name: "Apple", price: 1.99, quantity: 1 }],
      },
      theme: {
        theme: "dark",
      },
    });

    const { container } = render(
      <Provider store={store}>
        <Basket />
      </Provider>
    );

    expect(container.firstChild).toHaveClass("styled-basket dark");
  });

  test("Basket component renders correctly and matches snapshot", () => {
    store = mockStore({
      basket: {
        active: true,
        items: [{ id: 1, name: "Apple", price: 1.99, quantity: 1 }],
      },
      theme: {
        theme: "light",
      },
    });

    const tree = render(
      <Provider store={store}>
        <Basket />
      </Provider>
    );

    expect(tree.asFragment()).toMatchSnapshot();
  });
});
