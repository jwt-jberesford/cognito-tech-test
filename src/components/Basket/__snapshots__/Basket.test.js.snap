// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basket component unit tests Basket component renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="styled-basket light"
  >
    <h3>
      Your Basket
    </h3>
    <div
      class="basket-items"
    >
      <div
        class="item-wrapper"
      >
        <h4
          class="item-name"
        >
          Apple
        </h4>
        <div
          class="item-right"
        >
          <p
            class="item-price"
          >
            Â£1.99
          </p>
          <div
            class="item-quantity"
          >
            <button
              class="prev-button"
              disabled=""
            >
              <div
                class="svg-container"
              >
                <svg
                  class="prev-icon icon-transition icon-light disabled"
                  fill="none"
                  height="54"
                  viewBox="0 0 31 54"
                  width="31"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M29.9357 52.9424C28.5216 54.3525 26.2416 54.3525 24.8274 52.9424L0.844174 29.0288C-0.281392 27.9065 -0.281392 26.0935 0.844174 24.9712L24.8274 1.05755C26.2416 -0.352518 28.5216 -0.352518 29.9357 1.05755C31.3499 2.46763 31.3499 4.74101 29.9357 6.15108L9.04061 27.0144L29.9646 47.8777C31.3499 49.259 31.3499 51.5612 29.9357 52.9424Z"
                  />
                </svg>
              </div>
            </button>
            <span>
              1
            </span>
            <button
              class="next-button"
            >
              <div
                class="svg-container"
              >
                <svg
                  class="next-icon icon-transition icon-light "
                  fill="none"
                  height="54"
                  viewBox="0 0 31 54"
                  width="31"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M1.06427 52.9424C2.47844 54.3525 4.75843 54.3525 6.1726 52.9424L30.1558 29.0288C31.2814 27.9065 31.2814 26.0935 30.1558 24.9712L6.1726 1.05755C4.75843 -0.352518 2.47844 -0.352518 1.06427 1.05755C-0.349908 2.46763 -0.349908 4.74101 1.06427 6.15108L21.9594 27.0144L1.0354 47.8777C-0.349908 49.259 -0.349908 51.5612 1.06427 52.9424Z"
                  />
                </svg>
              </div>
            </button>
          </div>
          <button
            class="remove-item light"
          >
            Remove
          </button>
        </div>
      </div>
      <div
        class="break-line"
      />
    </div>
    <div
      class="checkout-button-wrapper"
    >
      <button
        class="primary button large light"
      >
        Go to checkout
      </button>
    </div>
  </div>
</DocumentFragment>
`;
